File: aspnet-in-practice/16/CS/Minifying-Markup/App_Code/CustomResponseModule.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Configuration;
using System.IO;
using System.Web.UI;

namespace ASPNET4InPractice.Chapter14
{
	public class CustomResponseModule : IHttpModule
	{
		public void Dispose()
		{
			// nothing to do
		}

		public void Init(HttpApplication context)
		{
			context.PreRequestHandlerExecute += new EventHandler(AddFilter);
		}

		void AddFilter(object sender, EventArgs e)
		{
			HttpApplication app = (HttpApplication)sender;
			
			// ignore non page and AJAX requests
			if (!(app.Context.CurrentHandler is Page) || !string.IsNullOrEmpty(app.Request["HTTP_X_MICROSOFTAJAX"]))
				return;

			// this is a bug in ASP.NET>3.5, you just need to first access Request.Filter
			Stream filter = app.Response.Filter;

			// and then add our custom Filter
			app.Response.Filter = new ResponseFilter(app.Response.OutputStream);
		}

	}

}

File: aspnet-in-practice/16/VB/Minifying-Markup/App_Code/CustomResponseModule.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Configuration
Imports System.IO
Imports System.Web.UI

Namespace ASPNET4InPractice.Chapter14
    Public Class CustomResponseModule
        Implements IHttpModule
        Public Sub Dispose() Implements IHttpModule.Dispose
            ' nothing to do
        End Sub

        Public Sub Init(ByVal context As HttpApplication) Implements IHttpModule.Init
            AddHandler context.PreRequestHandlerExecute, AddressOf AddFilter
        End Sub

        Private Sub AddFilter(ByVal sender As Object, ByVal e As EventArgs)
            Dim app As HttpApplication = DirectCast(sender, HttpApplication)

            ' ignore non page and AJAX requests
            If Not (TypeOf app.Context.CurrentHandler Is Page) OrElse Not String.IsNullOrEmpty(app.Request("HTTP_X_MICROSOFTAJAX")) Then
                Exit Sub
            End If

            ' this is a bug in ASP.NET>3.5, you just need to first access Request.Filter
            Dim filter As Stream = app.Response.Filter

            ' and then add our custom Filter
            app.Response.Filter = New ResponseFilter(app.Response.OutputStream)
        End Sub

    End Class

End Namespace